src/parser_urdf.cc

```diff
@@ -2663,48 +2665,137 @@ void URDF2SDF::ListSDFExtensions(const std::string &_reference)
 void CreateSDF(tinyxml2::XMLElement *_root,
                urdf::LinkConstSharedPtr _link)
 {
-  // must have an <inertial> block and cannot have zero mass.
-  //  allow det(I) == zero, in the case of point mass geoms.
+  // Links without an <inertial> block will be considered to have zero mass.
+  const bool linkHasZeroMass = !_link->inertial || _link->inertial->mass <= 0;
+
+  // link must have an <inertial> block and cannot have zero mass, if its parent
+  // joint and none of its child joints are reduced.
+  // allow det(I) == zero, in the case of point mass geoms.
   // @todo:  keyword "world" should be a constant defined somewhere else
-  if (_link->name != "world" &&
-      ((!_link->inertial) || gz::math::equal(_link->inertial->mass, 0.0)))
+  if (_link->name != "world" && linkHasZeroMass)
   {
-    if (!_link->child_links.empty())
+    Errors nonJointReductionErrors;
+    std::stringstream errorStream;
+    errorStream << "urdf2sdf: link[" << _link->name << "] has ";
+    if (!_link->inertial)
+    {
+      errorStream << "no <inertial> block defined. ";
+    }
+    else
     {
-      sdfdbg << "urdf2sdf: link[" << _link->name
-             << "] has no inertia, ["
-             << static_cast<int>(_link->child_links.size())
-             << "] children links ignored.\n";
+      errorStream << "a mass value of less than or equal to zero. ";
     }
+    errorStream << "Please ensure this link has a valid mass to prevent any "
+                << "missing links or joints in the resulting SDF model.";
+    nonJointReductionErrors.emplace_back(
+        ErrorCode::LINK_INERTIA_INVALID, errorStream.str());
+    errorStream.str(std::string());
 
-    if (!_link->child_joints.empty())
+    // If joint reduction happens and resolves the issue of this link with no
+    // mass, no warnings will be emitted
+    bool jointReductionHappens = _link->parent_joint &&
+        FixedJointShouldBeReduced(_link->parent_joint) &&
+        g_reduceFixedJoints;
+
+    // check if parent joint can be lumped and reduced
+    if (!jointReductionHappens && _link->parent_joint)
     {
-      sdfdbg << "urdf2sdf: link[" << _link->name
-             << "] has no inertia, ["
-             << static_cast<int>(_link->child_links.size())
-             << "] children joints ignored.\n";
+      if (_link->parent_joint->type == urdf::Joint::FIXED &&
+          (!FixedJointShouldBeReduced(_link->parent_joint) ||
+              !g_reduceFixedJoints))
+      {
```

Comment:
Do we need to check this again. Is this not equivalent to `!jointReductionHappens`, which is checked in the `if` statement right above?

Reply:
Thanks for catching that messy statement, we'll still need to check if the joint is fixed, as `!jointReductionHappens` could be due to the gazebo tags, I've cleaned it up in https://github.com/gazebosim/sdformat/pull/1238/commits/bb9fa6367488c3fac01be992f2347e4998f404d2